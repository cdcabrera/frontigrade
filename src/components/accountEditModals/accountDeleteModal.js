import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, Grid, Icon, Modal, Spinner } from 'patternfly-react';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';
import { fieldValidation } from '../formField/formField';

const initialState = {
  accountId: null,
  displayName: ''
};

class AccountDeleteModal extends React.Component {
  state = {
    ...initialState
  };

  onCancel = () => {
    this.setState(
      {
        ...initialState
      },
      () => {
        store.dispatch({
          type: reduxTypes.account.DELETE_ACCOUNT_HIDE
        });
      }
    );
  };

  onSubmit = event => {
    const { accountId, displayName } = this.state;
    const { deleteAccount } = this.props;

    event.preventDefault();

    deleteAccount(accountId).then(
      () => {
        this.onCancel();

        store.dispatch({
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: 'success',
          message: (
            <span>
              Account <strong>{displayName}</strong> successfully deleted.
            </span>
          )
        });

        store.dispatch({
          type: reduxTypes.account.UPDATE_ACCOUNTS
        });
      },
      () => {
        // eslint-disable-next-line react/destructuring-assignment
        if (Math.floor(this.props.errorStatus / 100) === 5) {
          this.onCancel();
          return;
        }

        // eslint-disable-next-line react/destructuring-assignment
        if (!this.props.show) {
          store.dispatch({
            type: reduxTypes.toastNotifications.TOAST_ADD,
            alertType: 'error',
            header: `Error Deleting Account ${displayName}`,
            message: `${this.props.errorMessage}` // eslint-disable-line react/destructuring-assignment
          });
        }
      }
    );
  };

  static getDerivedStateFromProps(props) {
    let updateInitialState = null;

    if (props.account && props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME]) {
      const accountId = props.account[apiTypes.API_RESPONSE_ACCOUNTS_ID];
      const accountName = props.account[apiTypes.API_RESPONSE_ACCOUNTS_NAME];
      const displayName = fieldValidation.isEmpty(accountName) ? accountId : accountName;

      updateInitialState = {
        accountId,
        displayName
      };
    }

    return updateInitialState;
  }

  renderErrorMessage() {
    const { error, errorMessage, pending } = this.props;

    if (error && !pending) {
      return (
        <Alert type="error" onDismiss={this.errorDismissed}>
          <strong>Error</strong>: {errorMessage}
        </Alert>
      );
    }

    return null;
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <React.Fragment>
          <Spinner loading size="lg" className="blank-slate-pf-icon" />
          <div className="text-center">Deleting account...</div>
        </React.Fragment>
      );
    }

    return null;
  }

  render() {
    const { displayName } = this.state;
    const { pending, show } = this.props;

    return (
      <Modal show={show} onHide={this.cancel}>
        <Modal.Header>
          <Button className="close" onClick={this.onCancel} aria-hidden="true" aria-label="Close">
            <Icon type="pf" name="close" />
          </Button>
          <Modal.Title>{`Delete Account - ${displayName}`}</Modal.Title>
        </Modal.Header>
        <Modal.Body />
        <Grid fluid>
          {this.renderPendingMessage()}
          {this.renderErrorMessage()}
          {!pending && <p>Are you sure you want to delete the account {displayName}?</p>}
        </Grid>
        <Modal.Footer>
          <Button bsStyle="default" className="btn-cancel" onClick={this.onCancel}>
            Cancel
          </Button>
          <Button bsStyle="primary" type="submit" onClick={this.onSubmit} disabled={pending}>
            Delete
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

AccountDeleteModal.propTypes = {
  account: PropTypes.object.isRequired, // eslint-disable-line react/no-unused-prop-types
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  errorStatus: PropTypes.number,
  pending: PropTypes.bool,
  show: PropTypes.bool.isRequired,
  deleteAccount: PropTypes.func
};

AccountDeleteModal.defaultProps = {
  error: false,
  errorMessage: null,
  errorStatus: null,
  pending: false,
  deleteAccount: helpers.noop
};

const mapDispatchToProps = dispatch => ({
  deleteAccount: id => dispatch(reduxActions.account.deleteAccount(id))
});

const mapStateToProps = state => ({ ...state.accountEdit.del });

const ConnectedAccountDeleteModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountDeleteModal);

export { ConnectedAccountDeleteModal as default, ConnectedAccountDeleteModal, AccountDeleteModal };
