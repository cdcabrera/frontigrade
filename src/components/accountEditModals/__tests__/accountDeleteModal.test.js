import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedAccountDeleteModal, AccountDeleteModal } from '../accountDeleteModal';

describe('AccountDeleteModal Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component with default props', () => {
    const store = generateEmptyStore({ accountEdit: { del: { show: true, account: {} } } });
    const component = shallow(<ConnectedAccountDeleteModal />, { context: { store } });

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {
      show: true,
      account: {}
    };

    const component = mount(<AccountDeleteModal {...props} />);
    expect(component.render()).toMatchSnapshot();
  });
});
