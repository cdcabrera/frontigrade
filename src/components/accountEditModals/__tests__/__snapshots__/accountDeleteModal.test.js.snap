// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDeleteModal Component should render a connected component with default props: connected 1`] = `
<AccountDeleteModal
  account={Object {}}
  deleteAccount={[Function]}
  error={false}
  errorMessage={null}
  errorStatus={null}
  pending={false}
  show={true}
/>
`;

exports[`AccountDeleteModal Component should render a non-connected component 1`] = `
<div
  role="dialog"
>
  <div
    class="fade modal-backdrop in"
  />
  <div
    class="fade in modal"
    role="dialog"
    style="display: block; padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
        role="document"
      >
        <div
          class="modal-header"
        >
          <button
            aria-hidden="true"
            aria-label="Close"
            class="close btn btn-default"
            type="button"
          >
            <span
              aria-hidden="true"
              class="pficon pficon-close"
            />
          </button>
          <h4
            class="modal-title"
          >
            Delete Account - 
          </h4>
        </div>
        <div
          class="modal-body"
        />
        <div
          class="container-fluid"
        >
          <p>
            Are you sure you want to delete the account ?
          </p>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn-cancel btn btn-default"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
